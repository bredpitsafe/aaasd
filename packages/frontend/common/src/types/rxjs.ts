import type { Observable, Subject, Subscription } from 'rxjs';

export interface IObserver<T, E> {
    next: (value: T) => void;
    error: (err: E) => void;
    complete: () => void;
}

export interface IObservable<T, E> extends Observable<T> {
    subscribe(next: (value: T) => void): Subscription;
    subscribe(observer?: Partial<IObserver<T, E>>): Subscription;
}

export interface ISubject<T, E> extends Subject<T> {
    error(err: E): void;
    subscribe(next: (value: T) => void): Subscription;
    subscribe(observer?: Partial<IObserver<T, E>>): Subscription;
}
