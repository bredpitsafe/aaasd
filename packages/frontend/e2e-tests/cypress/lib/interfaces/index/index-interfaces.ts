export enum EPagesName {
    dashboard = 'Dashboard',
    tradingServersManager = 'Trading Servers Manager',
    herodotusTerminal = 'Herodotus Terminal',
    herodotusTrades = 'Herodotus Trades',
    herodotusDashboard = 'Herodotus Dashboard',
    tradingStats = 'Trading Stats',
    backtestingManager = 'Backtesting Manager',
    jsonCharts = 'JSON Charts',
    balanceMonitor = 'Balance Monitor',
    wsQueryTerminal = 'WS Query Terminal',
}
