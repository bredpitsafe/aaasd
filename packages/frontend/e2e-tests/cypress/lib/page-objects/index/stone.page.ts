import { BasePage } from '../../base.page';
import { Button } from '../../base/elements/button';
import { Text } from '../../base/elements/text';
enum EStonePageSelectors {
    Dashboard = '[href*=dashboard]',
    TradingServersManager = '[href*=trading-servers-manager]',
    HerodotusTerminal = '[href*=herodotus-terminal]',
    TradingStats = '[href*=trading-stats]',
    BacktestingManager = '[href*=backtesting]',
    BalanceMonitor = '[href*=balance-monitor]',
    WSQueryTerminal = '[href*=ws-query-terminal]',
    JsonCharts = '[href*=charter-json-viewer]',
    LoadingIcon = '[class*="anticon anticon-loading"]',
}

class StonePage extends BasePage {
    readonly dashboard = new Text(EStonePageSelectors.Dashboard, false);
    readonly tradingServersManager = new Text(EStonePageSelectors.TradingServersManager, false);
    readonly herodotusTerminal = new Text(EStonePageSelectors.HerodotusTerminal, false);
    readonly tradingStats = new Text(EStonePageSelectors.TradingStats, false);
    readonly backtestingManager = new Text(EStonePageSelectors.BacktestingManager, false);
    readonly balanceMonitor = new Text(EStonePageSelectors.BalanceMonitor, false);
    readonly wsQueryTerminal = new Text(EStonePageSelectors.WSQueryTerminal, false);
    readonly jsonCharts = new Text(EStonePageSelectors.JsonCharts, false);
    readonly loadingIcon = new Button(EStonePageSelectors.LoadingIcon, false);

    checkElementsExists(): void {
        this.dashboard.checkExists();
        this.tradingServersManager.checkExists();
        this.herodotusTerminal.checkExists();
        this.tradingStats.checkExists();
        this.backtestingManager.checkExists();
        this.backtestingManager.checkExists();
    }

    open(): StonePage {
        this.visit();
        return this;
    }
}

export const stonePage = new StonePage();
