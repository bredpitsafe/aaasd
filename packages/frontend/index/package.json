{
  "name": "@frontend/index",
  "version": "1.0.0",
  "description": "simple index page for choose application",
  "scripts": {
    "build": "NODE_ENV=production npx webpack",
    "lint": "eslint './src/**/*.{js,ts,tsx}' --quiet",
    "server": "npx webpack serve",
    "test_unit": "jest --passWithNoTests",
    "types": "tsc --noEmit"
  },
  "dependencies": {
    "@frontend/common": "^1.0.1",
    "@common/types": "^1.0.0"
  }
}
