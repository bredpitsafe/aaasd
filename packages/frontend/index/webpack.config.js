const {
    buildConfig,
    htmlConfig,
    defaultInOut,
    defaultDevServerConfig,
} = require('../../../webpack/configs');
const path = require('path');

const app = 'index';
const resolvePath = () => path.resolve(__dirname);

module.exports = buildConfig([
    defaultInOut([app], resolvePath),
    htmlConfig([app], resolvePath),
    defaultDevServerConfig(app),
]);
