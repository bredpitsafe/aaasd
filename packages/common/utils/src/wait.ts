import { firstValueFrom, interval, map } from 'rxjs';

export const wait = (ms = 100): Promise<void> =>
    firstValueFrom<void>(interval(ms).pipe(map(() => undefined)));
