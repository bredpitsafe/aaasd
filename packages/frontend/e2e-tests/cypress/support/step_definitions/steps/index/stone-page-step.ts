import { Given } from '@badeball/cypress-cucumber-preprocessor';

import { EPagesName } from '../../../../lib/interfaces/index/index-interfaces';
import { ERobots } from '../../../../lib/interfaces/trading-servers-manager/robots-interfaces';
import { backtestingPage } from '../../../../lib/page-objects/backtesting/backtesting.page';
import { balanceMonitorPage } from '../../../../lib/page-objects/balance-monitor/balance-monitor.page';
import { charterJsonViewerPage } from '../../../../lib/page-objects/charter-json-viewer/charter-json-viewer.page';
import { authenticationPage } from '../../../../lib/page-objects/common/authentication.page';
import { settingsModal } from '../../../../lib/page-objects/common/settings.modal';
import { tableRow } from '../../../../lib/page-objects/common/table/table.row';
import { dashboardPage } from '../../../../lib/page-objects/dashboard/dashboard.page';
import { herodotusTerminalPage } from '../../../../lib/page-objects/herodotus-terminal/herodotus-terminal.page';
import { herodotusTradesPage } from '../../../../lib/page-objects/herodotus-trades/herodotus-trades.page';
import { stonePage } from '../../../../lib/page-objects/index/stone.page';
import { tradingServersManagerPage } from '../../../../lib/page-objects/trading-servers-manager/trading-servers-manager.page';
import { tradingStatsPage } from '../../../../lib/page-objects/trading-stats/trading-stats.page';
import { wsQueryTerminalPage } from '../../../../lib/page-objects/ws-query-terminal/ws-query-terminal.page';
import { customWait } from '../../../../lib/web-socket/server';
import { checkUrlInclude } from '../../../asserts/comon-url-assert';

Given(`user goes to the "Stone" page`, () => {
    stonePage.visit();
    stonePage.checkElementsExists();
});

Given(`user click an the {string} link`, (pageName: EPagesName) => {
    switch (pageName) {
        case EPagesName.dashboard:
            stonePage.dashboard.click();
            break;
        case EPagesName.tradingServersManager:
            stonePage.tradingServersManager.click();
            break;
        case EPagesName.herodotusTerminal:
            stonePage.herodotusTerminal.click();
            break;
        case EPagesName.tradingStats:
            stonePage.tradingStats.click();
            break;
        case EPagesName.backtestingManager:
            stonePage.backtestingManager.click();
            break;
        case EPagesName.jsonCharts:
            stonePage.jsonCharts.click();
            break;
        case EPagesName.balanceMonitor:
            stonePage.balanceMonitor.click();
            break;
        case EPagesName.wsQueryTerminal:
            stonePage.wsQueryTerminal.click();
            break;
    }
});

Given(
    `user on the {string} page with the "default" backend server parameter`,
    (pageName: EPagesName) => {
        switch (pageName) {
            case EPagesName.tradingServersManager:
                tradingServersManagerPage.checkElementsExists();
                break;
            case EPagesName.herodotusTerminal:
                herodotusTerminalPage.checkElementsExists();
                break;
            case EPagesName.tradingStats:
                tradingStatsPage.checkElementsExists();
                break;
            case EPagesName.backtestingManager:
                backtestingPage.checkElementsExists();
                break;
            case EPagesName.balanceMonitor:
                balanceMonitorPage.checkElementsExists();
                break;
        }
    },
);

Given(`user not sees data on the {string} page`, (pageName: EPagesName) => {
    switch (pageName) {
        case EPagesName.tradingServersManager:
            tradingServersManagerPage.checkNotVisibleTabPanel();
            tradingServersManagerPage.checkVisibleLoading();
            break;
        case EPagesName.herodotusTerminal:
            herodotusTerminalPage.checkVisibleTabPanel();
            herodotusTerminalPage.checkTaskTableVisibility(false);
            stonePage.loadingIcon.checkVisible();
            break;
        case EPagesName.backtestingManager:
            backtestingPage.checkNotExistsTask();
            break;
    }
});

Given(`user sees the {string} page`, (pageName: EPagesName) => {
    switch (pageName) {
        case EPagesName.tradingServersManager:
            tradingServersManagerPage.checkVisibleTabPanel();
            checkUrlInclude(tradingServersManagerPage.pageUrl);
            break;
        case EPagesName.herodotusTerminal:
            herodotusTerminalPage.checkVisibleTabPanel();
            herodotusTerminalPage.checkTaskTableVisibility(true);
            checkUrlInclude(herodotusTerminalPage.pageUrl);
            break;
        case EPagesName.tradingStats:
            tradingStatsPage.checkDailyVisibleTable();
            checkUrlInclude(tradingStatsPage.pageUrl);
            break;
        case EPagesName.backtestingManager:
            backtestingPage.checkVisiblePanel();
            checkUrlInclude(backtestingPage.pageUrl);
            break;
        case EPagesName.balanceMonitor:
            balanceMonitorPage.checkVisiblePanel();
            checkUrlInclude(balanceMonitorPage.pageUrl);
            break;
        case EPagesName.dashboard:
            dashboardPage.checkElementsExists();
            checkUrlInclude(dashboardPage.pageUrl);
            break;
        case EPagesName.jsonCharts:
            charterJsonViewerPage.checkElementsExists();
            checkUrlInclude(charterJsonViewerPage.pageUrl);
            break;
        case EPagesName.herodotusTrades:
            herodotusTradesPage.checkElementsExists();
            herodotusTradesPage.checkVisibleTable();
            checkUrlInclude(herodotusTradesPage.pageUrl);
            break;
        case EPagesName.wsQueryTerminal:
            wsQueryTerminalPage.checkElementsExists();
            wsQueryTerminalPage.checkVisibleTab();
            checkUrlInclude(wsQueryTerminalPage.pageUrl);
            break;
    }
});

Given(
    `user goes to the {string} page with the backend server parameter`,
    (pageName: EPagesName) => {
        switch (pageName) {
            case EPagesName.tradingServersManager:
                tradingServersManagerPage.openPageWithBackendServer();
                break;
            case EPagesName.herodotusTerminal:
                herodotusTerminalPage.openPageWithBackendServer();
                break;
            case EPagesName.tradingStats:
                tradingStatsPage.openPageWithBackendServer();
                break;
            case EPagesName.backtestingManager:
                backtestingPage.openPageWithBackendServer();
                break;
            case EPagesName.balanceMonitor:
                balanceMonitorPage.openPageWithBackendServer();
                break;
            case EPagesName.wsQueryTerminal:
                wsQueryTerminalPage.openPageWithBackendServer();
                break;
        }
    },
);

Given(
    `user goes to the {string} page by {string} server params`,
    (pageName: EPagesName, nameServer: string) => {
        switch (pageName) {
            case EPagesName.tradingServersManager:
                tradingServersManagerPage.openPageByBackendServer(nameServer);
                break;
            case EPagesName.herodotusTerminal:
                herodotusTerminalPage.openPageByBackendServer(nameServer);
                break;
            case EPagesName.tradingStats:
                tradingStatsPage.openPageByBackendServer(nameServer);
                break;
            case EPagesName.backtestingManager:
                backtestingPage.openPageByBackendServer(nameServer);
                break;
            case EPagesName.balanceMonitor:
                balanceMonitorPage.openPageByBackendServer(nameServer);
                break;
        }
    },
);

Given(
    `user goes to the "Backtesting Manager" page with the backend server {string}`,
    (nameServer: string) => {
        backtestingPage.openPageWithBackendServerByParams(nameServer);
    },
);

Given(`user goes to the "Herodotus Terminal" page by index {string}`, (index: string) => {
    herodotusTerminalPage.openPageWithBackendServerByIndex(index);
});

Given(`user goes to the "Herodotus Terminal" page by name {string}`, (robotName: string) => {
    const index = `${ERobots[robotName]}`;
    herodotusTerminalPage.openPageWithBackendServerByIndex(index);
});

Given(`user click "Log Out" button on the {string} page`, () => {
    authenticationPage.clickLogOutButton();
});

Given(`user on the {string} page with the "ModalSettings" modal`, (pageName: EPagesName) => {
    settingsModal.checkElementsExists();
    customWait(1);
    switch (pageName) {
        case EPagesName.tradingServersManager:
            checkUrlInclude(tradingServersManagerPage.pageUrl);
            break;
        case EPagesName.herodotusTerminal:
            checkUrlInclude(herodotusTerminalPage.pageUrl);
            break;
        case EPagesName.tradingStats:
            checkUrlInclude(tradingStatsPage.pageUrl);
            break;
        case EPagesName.backtestingManager:
            checkUrlInclude(backtestingPage.pageUrl);
            break;
    }
});

Given(`user opens the {string} page`, (pageName: EPagesName) => {
    switch (pageName) {
        case EPagesName.dashboard:
            dashboardPage.visit();
            dashboardPage.checkElementsExists();
            break;
        case EPagesName.tradingServersManager:
            tradingServersManagerPage.visit();
            settingsModal.checkElementsExists();
            break;
        case EPagesName.herodotusTerminal:
            herodotusTerminalPage.visit();
            settingsModal.checkElementsExists();
            break;
        case EPagesName.tradingStats:
            tradingStatsPage.visit();
            settingsModal.checkElementsExists();
            break;
        case EPagesName.backtestingManager:
            backtestingPage.visit();
            settingsModal.checkElementsExists();
            break;
        case EPagesName.balanceMonitor:
            balanceMonitorPage.visit();
            settingsModal.checkElementsExists();
            break;
        case EPagesName.wsQueryTerminal:
            wsQueryTerminalPage.visit();
            settingsModal.checkElementsExists();
            break;
    }
});

Given(`user on the "Herodotus Terminal" page of the {string} robot`, (robotName: ERobots) => {
    checkUrlInclude(herodotusTerminalPage.pageUrl);
    herodotusTerminalPage.checkElementsExists();
    checkUrlInclude(`${ERobots[robotName]}`);
});

Given(
    `user types the modal settings with the {string} backend server parameter`,
    (backendServerName: string) => {
        if (backendServerName === 'default') {
            backendServerName = Cypress.env('backendServerName');
        }
        settingsModal.serverSelect.checkVisible();
        customWait(0.5);
        settingsModal.serverSelect.typeAndClickByText(backendServerName);
    },
);

Given(`user gets URL the "Charts" page and opens the page`, () => {
    tableRow.visitPageByLinkInRow();
});
