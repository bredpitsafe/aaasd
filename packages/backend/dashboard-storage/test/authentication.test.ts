import { wait } from '@common/utils/src/wait.js';
import { firstValueFrom, skip } from 'rxjs';
import { afterAll, describe, expect, it, vi } from 'vitest';

import { getToken } from './client/authentication.js';
import { Client, users } from './client/index.js';

describe('Authentication', () => {
    const client = new Client();

    afterAll(() => {
        client.close();
    });

    it('successful', async () => {
        const token = await getToken(users.user1.username);
        const response = await client.request({ type: 'Authenticate', bearerToken: token });
        expect(response.error).toBeUndefined();
        expect(response.state).toBe('Done');
        expect(response.payload).toMatchSnapshot();
    });

    it('failed (no user)', async () => {
        const token = await getToken('unknown_user');
        const response = await client.request({ type: 'Authenticate', bearerToken: token });
        expect(response.error).not.toBeUndefined();

        expect(response.error).toMatchSnapshot();
    });

    it('failed (no token)', async () => {
        const response = await client.request({ type: 'Authenticate', bearerToken: undefined });
        expect(response.error).not.toBeUndefined();

        expect(response.error).toMatchSnapshot();
    });

    it('failed (unknown token)', async () => {
        const response = await client.request({
            type: 'Authenticate',
            bearerToken: 'unknown_token',
        });
        expect(response.error).not.toBeUndefined();

        expect(response.error).toMatchSnapshot();
    });

    it('failed (expired token)', async () => {
        const token = await getToken(users.short_exp.username);
        await wait(users.short_exp.exp * 1000);
        const response = await client.request({ type: 'Authenticate', bearerToken: token });

        expect(response.error).toMatchSnapshot();
    });

    it('authentication expires with an active subscription', async () => {
        const authenticatedClient = new Client({
            authenticate: true,
            as: users.short_exp.username,
        });
        const response = await firstValueFrom(
            authenticatedClient.send({ type: 'SubscribeToDashboardsList' }).pipe(skip(1)),
        );

        expect(response.error).toMatchSnapshot();
        authenticatedClient.close();
    });

    it('logout (non-authenticated socket)', async () => {
        const response = await client.request({ type: 'Logout' });
        expect(response).toMatchObject({
            state: 'Done',
            payload: {
                type: 'LoggedOut',
            },
        });
    });

    it('logout stops active subscription', async () => {
        const authenticatedClient = new Client({ authenticate: true });
        const subscriber = vi.fn();
        const subscription = authenticatedClient
            .send({ type: 'SubscribeToDashboardsList' })
            .subscribe(subscriber);

        await wait();
        const response = await authenticatedClient.request({ type: 'Logout' });
        expect(response).toMatchObject({
            state: 'Done',
            payload: {
                type: 'LoggedOut',
            },
        });

        await wait(1000);

        expect(subscriber.mock.calls).toHaveLength(2);
        const lastResponse = subscriber.mock.calls[1][0];
        expect(lastResponse.error).toMatchSnapshot();

        subscription.unsubscribe();
    });

    it('logout closes socket', async () => {
        const authenticatedClient = new Client({ authenticate: true });
        const subscription = authenticatedClient
            .send({ type: 'SubscribeToDashboardsList' })
            .subscribe();

        await wait(1000);
        await authenticatedClient.request({ type: 'Logout' });
        await wait(1100);

        expect(subscription.closed).toBe(true);
    });

    it('authentication expiration closes socket with subscription error', async () => {
        const authenticatedClient = new Client({
            authenticate: true,
            as: users.short_exp.username,
        });
        const subscriber = vi.fn();
        const subscription = authenticatedClient
            .send({ type: 'SubscribeToDashboardsList' })
            .subscribe(subscriber);

        await wait(users.short_exp.exp * 1000 + 100);

        expect(subscriber.mock.calls).toHaveLength(2);
        expect(subscriber.mock.calls[1][0].error).not.toBeUndefined();
        expect(subscriber.mock.calls[1][0].error).toMatchSnapshot();

        await wait(1000);

        expect(subscription.closed).toBe(true);
    });
});
