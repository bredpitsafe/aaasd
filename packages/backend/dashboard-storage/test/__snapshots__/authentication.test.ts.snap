// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Authentication > authentication expiration closes socket with subscription error 1`] = `
{
  "description": "AuthenticationRequired: Bearer token was not provided or has expired in the current socket",
  "kind": "Authentication",
}
`;

exports[`Authentication > authentication expires with an active subscription 1`] = `
{
  "description": "AuthenticationRequired: Bearer token was not provided or has expired in the current socket",
  "kind": "Authentication",
}
`;

exports[`Authentication > failed (expired token) 1`] = `
{
  "description": "Authentication Failed: "exp" claim timestamp check failed",
  "kind": "Authentication",
}
`;

exports[`Authentication > failed (no token) 1`] = `
{
  "description": "Request validation error: request body is missing required fields since it does not match any request schema.",
  "kind": "Validation",
}
`;

exports[`Authentication > failed (no user) 1`] = `
{
  "description": "Authentication Failed: "exp" claim timestamp check failed",
  "kind": "Authentication",
}
`;

exports[`Authentication > failed (unknown token) 1`] = `
{
  "description": "Authentication Failed: Invalid JWT",
  "kind": "Authentication",
}
`;

exports[`Authentication > logout stops active subscription 1`] = `
{
  "description": "AuthenticationRequired: Bearer token was not provided or has expired in the current socket",
  "kind": "Authentication",
}
`;

exports[`Authentication > successful 1`] = `
{
  "type": "Authenticated",
  "username": "user1",
}
`;
